### PR TITLE
Add substack cmd for mathtext

### LINKED ISSUE #14247: latex \substack does't work
### Bug report

**Bug summary**

**Code for reproduction**

```python
# Paste your code here
#
#
```

**Actual outcome**

```
# If applicable, paste the console output here
#
#
```

**Expected outcome**

**Matplotlib version**

  * Operating system: win10
  * Matplotlib version: 3.0.1
  * Matplotlib backend (`print(matplotlib.get_backend())`):
  * Python version: python37
  * Jupyter version (if applicable):
  * Other libraries:

### PR DESCRIPTION
**Pull request by devRD**
## PR summary
Fixes #14247 

## PR checklist

- [x] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [ ] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [ ] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/documenting_mpl.html#writing-examples-and-tutorials)
- [ ] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/coding_guide.html#new-features-and-api-changes)
- [ ] Documentation complies with [general](https://matplotlib.org/devdocs/devel/documenting_mpl.html#writing-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/documenting_mpl.html#writing-docstrings) guidelines

### CODE REVIEW

File: lib/matplotlib/tests/test_mathtext.py [hunk 1]
```diff
@@ -134,7 +134,7 @@
     r'$\sum x\quad\sum^nx\quad\sum_nx\quad\sum_n^nx\quad\prod x\quad\prod^nx\quad\prod_nx\quad\prod_n^nx$',  # GitHub issue 18085
     r'$1.$ $2.$ $19680801.$ $a.$ $b.$ $mpl.$',
     r'$\text{text}_{\text{sub}}^{\text{sup}} + \text{\$foo\$} + \frac{\text{num}}{\mathbf{\text{den}}}\text{with space, curly brackets \{\}, and dash -}$',
-
+    r'$\sum_{\substack{k = 1}{k \neq \lfloor n/2\rfloor}}^{n}P(i,j) \sum_{\substack{i \neq 0}{-1 \leq i \leq 3}{1 \leq j \leq 5}} F^i(x,y)$',
```
Comment (oscargus) [inline lib/matplotlib/tests/test_mathtext.py:137]: Although I believe that your syntax makes very much sense, it appears as if the existing `\substack` command is simply using `\\` for next line.

```suggestion
    r'$\sum_{\substack{k = 1\\ k \neq \lfloor n/2\rfloor}}^{n}P(i,j) \sum_{\substack{i \neq 0\\ -1 \leq i \leq 3\\ 1 \leq j \leq 5}} F^i(x,y)$',
```

Apart from that, the outcome looks good! Maybe one would like to test with `\left \lfloor \frac{n}{2} \right\rfloor` as well to see how that behaves?

File: lib/matplotlib/_mathtext.py [hunk 2]
```diff
@@ -2599,3 +2604,23 @@ def auto_delim(self, s, loc, toks):
             # if "mid" in toks ... can be removed when requiring pyparsing 3.
             toks["mid"].asList() if "mid" in toks else [],
             toks["right"])
+
+    def substack(self, s, loc, toks):
+        state = self.get_state()
+        thickness = state.get_current_underline_thickness()
+        vlist = []
+
+        max_width = sorted(toks[1:], key= lambda c: c.width)[-1].width
```
Comment (tfpf) [inline lib/matplotlib/_mathtext.py:2613]: Maybe this could be written using `max`, which also takes a `key` argument.
```
max_width = max(toks[1:], key=lambda c: c.width)
```
Something like that?

File: lib/matplotlib/tests/test_mathtext.py (hunk 1 above)
Comment (devRD) [reply lib/matplotlib/tests/test_mathtext.py]: Not really sure how to go about adding the `\\` as next-line behavior to the `substack` input due to the python's handling of escape characters, I'm currently still looking into it.

Added the `\left \lfloor \frac{n}{2} \right\rfloor` as a test with the current implemention.

Comment (oscargus) [reply lib/matplotlib/tests/test_mathtext.py]: Maybe the parser expert @anntzer has some input?

File: lib/matplotlib/_mathtext.py [hunk 3]
```diff
@@ -2599,3 +2604,23 @@ def auto_delim(self, s, loc, toks):
...
+
+        for sub in toks[1:]:
+            cp = HCentered([sub])
+            cp.hpack(max_width, 'exactly')
+            vlist.append(cp)
+
+        vlist = [x for l in (vlist[i:i+1] +
+                             [Vbox(0, thickness * 2.0)] *
+                             (i < len(vlist) - 1)
+                             for i in range(0, len(vlist), 1)) for x in l]
```
Comment (tfpf) [inline lib/matplotlib/_mathtext.py:2623]: Took me an embarrassingly long time to understand!:sweat_smile: I think it separates each element of `vlist` with `Vbox(0, thickness * 2)`?

This is probably one of those instances I'd prioritise readability.
```
vlist_ = [None, Vbox(0, thickness * 2)] * len(vlist)
vlist_[::2] = vlist
vlist_.pop()
```
This should achieve this same thing. (Another option is to append the `Vbox`es in the `for` loop just above.)

However, I'd defer to @oscargus's opinion: what do you think?

File: lib/matplotlib/tests/test_mathtext.py (hunk 1 above)
Comment (ksunden) [reply lib/matplotlib/tests/test_mathtext.py]: I would expect pyparsing's [`DelimitedList`](https://pyparsing-docs.readthedocs.io/en/latest/pyparsing.html#pyparsing.DelimitedList) would be helpful here

Note, however, that it looks like it is in a slightly awkward transition from the `delimited_list` function to the `DelimitedList` class, the former being deprecated upon the introduction of the latter in pyparsing 3.1.0 (which we are incompatible with for other reasons right now, but should fix)

As such may need something like:

```python
if pyparsing.__version__info__ < (3, 1):
    from pyparsing import delimited_list as DelimitedList
else:
    from pyparsing import DelimitedList
```

File: lib/matplotlib/_mathtext.py (hunk 3 above)
Comment (oscargus) [reply lib/matplotlib/_mathtext.py:2623]: Maybe one could just do:

```
vlist = [val for pair in zip(vlist, [Vbox(0, thickness * 2)]*len(vlist)) for val in pair]
del vlist[-1]
```
?

(But, yes, I find double (and triple) list comprehensions quite hard to follow to start with.)

**Review by oscargus — APPROVED** [review]

File: lib/matplotlib/_mathtext.py [hunk 4]
```diff
@@ -27,6 +27,11 @@
 from .font_manager import FontProperties, findfont, get_font
 from .ft2font import FT2Image, KERNING_DEFAULT
 
+from pyparsing import __version__ as pyparsing_version
+if pyparsing_version < '3.0.0':
```
Comment (ksunden) [inline lib/matplotlib/_mathtext.py:31]: This falls into the category of lurking problems that _probably_ won't be a real problem but _could_ be...

string comparison of version numbers is likely to fail. 

Unfortunately the version we are gaurding against doesn't have a tuple version (with integers) to compare to. (Current pyparsing _does_)

we are saved a bit by this having minor/patch version of 0, but would still fail if pyparsing got to version `10.0.0` as we would want that to be greater than 3.0.0, but string comparison would fail. Now _probably_ by the time that happens (_if_ it happens) we will no longer support versions less than 3... but if pyparsing decided to switch to date versions of `2023.7.0` or something, we would be in the same boat a whole lot sooner (and possibly without warning).

```suggestion
if tuple(int(x) for x in pyparsing_version.split(".")) < (3, 0, 0):
```

may work in a bit more robust manner

Or we can just use `nestedExpr` for now and handle switching to the new name later (as at least oneOf is already using the deprecated version, so that change can be separated from this PR.

**Review by ksunden — APPROVED** [review]
